### PR TITLE
Support for fontc and crater

### PR DESCRIPTION
**Pull request by cmyr**
Here is my attempt to add support in `gftools builder` for running `fontc`, and generally making it easier to use in the context of `fontc_crater`.

- adds the `--experimental-fontc` flag, which, if set:
    - replaces the `BuildVariable`, `BuildTTF`, and `BuildOTF` operations with alternatives that use `fontc`
    - modifies the config so that only _one_ of `buildVariable` and `buildStatic` is `true` (so do not instance variable fonts) and also so that only one of `buildTTF` and `buildOTF` are true (fontc does not currently build CFF outlines, so we turn `buildOTF` off if `buildTTF` is true)
- adds the `--experimental-simple-output=PATH` option, which:
    - if modifies the config so that all built fonts end up in one directory
    - and does the same config pruning as above, so that we can get the same set of outputs with both compilers.

Overall I've tried to make this as noninvasive as possible to the existing code, and I think that has been reasonably successful?

### CODE REVIEW

File: Lib/gftools/builder/operations/__init__.py
Comment (rsheeter) [inline Lib/gftools/builder/operations/__init__.py:175]: Operation registry is basically a factory. If we had a different factory for fontc vs dynamicly loaded python, created based on flags, then only the python one would need to do the dynamic module import bit. 

Considering most things use the googlefonts implementation and it always makes the same choices I find myself wondering if there should be a static implementation of that but I suppose that's out of scope.

Comment (cmyr) [reply Lib/gftools/builder/operations/__init__.py:175]: I presume the dynamic loading stuff exists for a reason, and I'm not trying to make any big changes here; just override some of the operations in the presence of a flag.

The rest of the operation types we're happy to share.

**Comment by cmyr** [comment]
Okay I've gone and cleaned this up to the point where I think it is ready to merge, or at least is ready for more considered review. Overall I'm quite happy with how this has gone; I've gotten more or less everything I was hoping for without needing to do anything very invasive.

File: Lib/gftools/builder/fontc.py
Comment (rsheeter) [inline Lib/gftools/builder/fontc.py:5]: Provide a link to the fontc repo, reader might not know what a fontc is.

File: Lib/gftools/builder/operations/fontc/__init__.py
Comment (rsheeter) [inline Lib/gftools/builder/operations/fontc/__init__.py:5]: _ prefix for a "private" thing.

Suggest default value of None.

File: Lib/gftools/builder/operations/fontc/__init__.py
Comment (rsheeter) [inline Lib/gftools/builder/operations/fontc/__init__.py:12]: This should fail if the FONTC_PATH is set already and fail if path doesn't point to a file. Maybe even fail if it doesn't point to something executable.

**Review by rsheeter — APPROVED** [review]

**Comment by rsheeter** [comment]
Heads up CI is upset with you

File: Lib/gftools/builder/operations/fontc/__init__.py
Comment (cmyr) [reply Lib/gftools/builder/operations/fontc/__init__.py:5]: I chose this default value so that if somehow this ended up not getting set there would be a slightly more informative error message. Is that too clever?

File: Lib/gftools/builder/operations/fontc/__init__.py
Comment (cmyr) [reply Lib/gftools/builder/operations/fontc/__init__.py:12]: I tried doing this logic and then python had some error about setting a global twice? so I think that's actually baked in now?

the check that the file exists happens earlier, at the call site.

Comment (rsheeter) [reply Lib/gftools/builder/operations/fontc/__init__.py:12]: Hm, I don't remember there being any rules about multiple writes to globals.

I just tried locally and python happily let me write to a global repeatedly.

**Comment by cmyr** [comment]
@m4rc1e would you like to take a look at this and let me know if it seems sane / if there's a simpler way I might achieve any of my goals, here?

File: Lib/gftools/builder/operations/__init__.py
Comment (anthrotype) [inline Lib/gftools/builder/operations/__init__.py:168]: I think you forgot a `return FontcBuildTTF`...

File: Lib/gftools/builder/operations/fontc/fontcBuildOTF.py
Comment (anthrotype) [inline Lib/gftools/builder/operations/fontc/fontcBuildOTF.py:4]:
```suggestion
class FontcBuildOTF(FontcOperationBase):
```

File: Lib/gftools/builder/file.py
Comment (m4rc1e) [inline Lib/gftools/builder/file.py:52]: This may be too simple since I have a seen a family that has 4 masters and 4 instances which are not MM compatible so it could only export 4 static fonts.

Something like `self.is_glyphs and len(self.gsfont.masters) < len(self.gsfont.instances)` but it still isn't 100% reliable.

File: Lib/gftools/builder/operations/fontc/__init__.py
Comment (m4rc1e) [inline Lib/gftools/builder/operations/fontc/__init__.py:31]: Nit. I'm probably being dense here but couldn't skip the reverse part and just keep popping the first element by using `args.pop(0)`,

File: Lib/gftools/builder/file.py
Comment (anthrotype) [reply Lib/gftools/builder/file.py:52]: perhaps `maybe_variable` is a better name for the method. But I think the gftools config provides an even stronger hint as to whether a font is supposed to be variable: if the font dev asks to build a variable font, then it must be variable; if it disables building the VF (is it `buildVariable: false`?) then one should treat it as not variable.

**Review by m4rc1e — COMMENTED** [review]
Nice work @cmyr. I'm not the original author of builder 2 so I'll let @simoncozens have a look before merging.

I've tested it on a few families and it's working well for its intended purposes.

I have no problems with this PR being merged. However, once fontc and its sibling libs/tools have matured, it would be great to drop the experimental flags and allow the user to specify which compiler they would like in the config file. This will drop the need to have a seperate `fontc_args` argument for the `GFBuilder` class, which should make the code a touch tidier.

My only slight hesitation is:

> modifies the config so that only one of buildVariable and buildStatic is true 

Currently, a config's default settings will export otf, ttf and var and make web versions for all formats. It's sad to see this go but since it is an experimental feature used solely by you folks, I'm ok with this.

**Review by m4rc1e — APPROVED** [review]

File: Lib/gftools/builder/file.py
Comment (simoncozens) [reply Lib/gftools/builder/file.py:52]: > But I think the gftools config provides an even stronger hint as to whether a font is supposed to be variable: if the font dev asks to build a variable font, then it must be variable; if they disables building the VF (is it buildVariable: false?) then one should treat it as not variable.

I'm not sure I understand this. `buildStatic`, `buildVariable`, `buildOTF`, `buildWebfont` etc. are parameters about *targets*, not about *sources*. "I don't want you to build a variable font today" does not mean "this is not a variable font".

**Pull request merged**
